### PR TITLE
Mode for `Round`

### LINKED ISSUE #3237: Add model for `round` build-in function
**Is your feature request related to a problem? Please describe.**
#3178 add feature that that specify the floating point accuracy.

We can further improve the feature by adding mode according to the [doc of neo4j](https://neo4j.com/docs/cypher-manual/current/functions/mathematical-numeric/#functions-round3) 

**Describe the solution you'd like**
You can refer to #3178.

What' more, you also refer to the source code of neo4j.

**Additional context**
None

**mentor**
@jackwener 

If you have any question, you can contact me.

### PR DESCRIPTION
**Pull request by Koukyosyumei**
## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### Issue(s) number: 

close #3237 

#### Description:

I added 7 modes implemented in neo4j.

<html>
<body>

UP | Round away from zero.
-- | --
DOWN | Round towards zero.
CEILING | Round towards positive infinity.
FLOOR | Round towards negative infinity.
HALF_UP | Round towards closest value of given precision, with ties always being rounded away from zero.
HALF_DOWN | Round towards closest value of given precision, with ties always being rounded towards zero.
HALF_EVEN | Round towards closest value of given precision, with ties always being rounded to the even neighbor.

</body>
</html>

## Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [x] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Comment by wey-gu** [comment]
Wow, welcome to the NebulaGraph community!! 🚀

Would you mind adding a `round.feature` e2e test under `tests/tck/features/function`?

**Review by wey-gu — DISMISSED** [review]
Thanks a lot for the clean PR!

**Comment by Koukyosyumei** [comment]
Thank you!! This project is really cool!!

**Comment by Koukyosyumei** [comment]
I fixed the format of `round.feature`

**Review by yixinglu — DISMISSED** [review]
interesting

File: src/common/function/FunctionManager.cpp [hunk 1]
```diff
@@ -68,8 +69,12 @@ std::unordered_map<std::string, std::vector<TypeSignature>> FunctionManager::typ
     {"round",
      {TypeSignature({Value::Type::INT}, Value::Type::FLOAT),
       TypeSignature({Value::Type::INT, Value::Type::INT}, Value::Type::FLOAT),
+      TypeSignature({Value::Type::FLOAT, Value::Type::INT, Value::Type::STRING},
+                    Value::Type::FLOAT),
       TypeSignature({Value::Type::FLOAT}, Value::Type::FLOAT),
-      TypeSignature({Value::Type::FLOAT, Value::Type::INT}, Value::Type::FLOAT)}},
+      TypeSignature({Value::Type::FLOAT, Value::Type::INT}, Value::Type::FLOAT),
+      TypeSignature({Value::Type::FLOAT, Value::Type::INT, Value::Type::STRING},
+                    Value::Type::FLOAT)}},
```
Comment (czpmango) [inline src/common/function/FunctionManager.cpp:77]: Duplicated?

File: src/common/function/FunctionManager.cpp [hunk 2]
```diff
@@ -577,10 +582,78 @@ FunctionManager::FunctionManager() {
         }
         case Value::Type::INT:
         case Value::Type::FLOAT: {
-          if (args.size() == 2) {
+          if (args.size() >= 2) {
             if (args[1].get().type() == Value::Type::INT) {
+              auto val = args[0].get().getFloat();
```
Comment (czpmango) [inline src/common/function/FunctionManager.cpp:587]: Add test case: `round(3, 2)`

Comment (czpmango) [reply src/common/function/FunctionManager.cpp:587]: Is args[0] always float type here?

File: src/common/function/FunctionManager.cpp [hunk 3]
```diff
@@ -577,10 +582,78 @@ FunctionManager::FunctionManager() {
...
               auto decimal = args[1].get().getInt();
-              return std::round(args[0].get().getFloat() * pow(10, decimal)) / pow(10, decimal);
+              auto factor = pow(10.0, decimal);
+
+              string mode = "half_up";
```
Comment (czpmango) [inline src/common/function/FunctionManager.cpp:591]: Default rounding mode here?
Add test case:
```
round(-1.5, 1)  
-1.0 is expected.
```

File: src/common/function/FunctionManager.cpp [hunk 4]
```diff
@@ -577,10 +582,78 @@ FunctionManager::FunctionManager() {
...
+              if (args.size() == 3 && args[2].get().type() == Value::Type::STRING) {
+                mode = args[2].get().getStr();
+              }
+              if (mode == "up") {
```
Comment (czpmango) [inline src/common/function/FunctionManager.cpp:595]: Make the rounding mode case insensitive.

File: tests/tck/features/function/round.feature [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+      """
+    Then the result should be, in any order:
+      | result |
+      | -1.3   |
+    When executing query:
+      """
+      RETURN round(1.25, 1, "half_even") as result
+      """
+    Then the result should be, in any order:
+      | result |
```
Comment (czpmango) [inline tests/tck/features/function/round.feature:206]: Add some test cases with nulls or illegal rounding mode.

File: tests/tck/features/function/round.feature [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+      | 1.2    |
+    When executing query:
+      """
+      RETURN round(-1.35, 1, "half_even") as result
```
Comment (czpmango) [inline tests/tck/features/function/round.feature:210]: Add some type mismatch cases:
```
round(3.125, 3.2)
```

File: src/common/function/test/FunctionManagerTest.cpp [hunk 7]
```diff
@@ -107,6 +107,34 @@ std::unordered_map<std::string, std::vector<Value>> FunctionManagerTest::args_ =
     {"round1", {11111.11111, 2}},
     {"round2", {11111.11111, -1}},
     {"round3", {11111.11111, -5}},
+    {"round4", {1.249, 1, "up"}},
```
Comment (czpmango) [inline src/common/function/test/FunctionManagerTest.cpp:110]: Test round(1.5, 0)

Comment (czpmango) [reply src/common/function/test/FunctionManagerTest.cpp:110]:
```
round(1.5, 0 "half_up") -> 2.0
round(1.5, 0 "half_down") -> 1.0
round(12.22, 0 "half_up") -> 12.0
```

File: src/common/function/FunctionManager.cpp [hunk 8]
```diff
@@ -577,10 +582,78 @@ FunctionManager::FunctionManager() {
...
+                  return next_largest;
+                } else {
+                  return next_smallest;
+                }
+              } else if (mode == "half_even") {
+                auto next_largest = std::ceil(val * factor) / factor;
+                auto next_smallest = std::floor(val * factor) / factor;
+                if (std::fabs((val - next_smallest) - (next_largest - val)) < FLT_EPSILON) {
+                  if (std::fmod(next_smallest * factor, 2.0) == 1.0) {
+                    return next_largest;
```
Comment (czpmango) [inline src/common/function/FunctionManager.cpp:645]: Too many if-else branch instructions may cause some performance issue, such as the icache miss, etc. Do you have any more efficient implementation scheme?

File: src/common/function/FunctionManager.cpp [hunk 9]
```diff
@@ -577,10 +582,78 @@ FunctionManager::FunctionManager() {
...
+              } else if (mode == "ceiling") {
+                return std::ceil(val * factor) / factor;
+              } else if (mode == "floor") {
+                return std::floor(val * factor) / factor;
+              } else if (mode == "half_up") {
+                return std::round(val * factor) / factor;
+              } else if (mode == "half_down") {
+                auto next_largest = std::ceil(val * factor) / factor;
+                auto next_smallest = std::floor(val * factor) / factor;
+                if (std::fabs((val - next_smallest) - (next_largest - val)) < FLT_EPSILON) {
```
Comment (czpmango) [inline src/common/function/FunctionManager.cpp:629]: There may be a more efficient way to determine tie.
Fwiw, static_cast<int>(val * factor)&0x0111.

File: tests/tck/features/function/round.feature (hunk 6 above)
Comment (czpmango) [reply tests/tck/features/function/round.feature:210]: round(3.123, -2)

**Comment by wey-gu** [comment]
Dear @Koukyosyumei ,

Just to ping you in case you missed the notification of the review :D, not pushy, kindly check @czpmango 's review comment when it's comfortable for you. Thanks!

BR//Wey

File: src/common/function/FunctionManager.cpp (hunk 3 above)
Comment (Koukyosyumei) [reply src/common/function/FunctionManager.cpp:591]: I guess `round(-1.5, 1) ` should be -1.5.

File: src/common/function/FunctionManager.cpp (hunk 9 above)
Comment (Koukyosyumei) [reply src/common/function/FunctionManager.cpp]: Thank you for your excellent suggestion! I have updated the logic not to calculate both of the largest and smallest.

**Comment by Koukyosyumei** [comment]
> Dear @Koukyosyumei ,
> 
> Just to ping you in case you missed the notification of the review :D, not pushy, kindly check @czpmango 's review comment when it's comfortable for you. Thanks!
> 
> BR//Wey

Sorry for the late response. I'm currently working on it!!

File: src/common/function/FunctionManager.cpp (hunk 8 above)
Comment (Koukyosyumei) [reply src/common/function/FunctionManager.cpp]: I concatenated some if-else branches.

File: src/common/function/test/FunctionManagerTest.cpp (hunk 7 above)
Comment (Koukyosyumei) [reply src/common/function/test/FunctionManagerTest.cpp:110]: I have added these cases.

File: src/common/function/FunctionManager.cpp (hunk 4 above)
Comment (Koukyosyumei) [reply src/common/function/FunctionManager.cpp]: I adopted boost:iequal to make the rounding mode comparison case insensitive.

File: src/common/function/FunctionManager.cpp (hunk 1 above)
Comment (Koukyosyumei) [reply src/common/function/FunctionManager.cpp:77]: I have fixed it.

File: src/common/function/FunctionManager.cpp (hunk 2 above)
Comment (Koukyosyumei) [reply src/common/function/FunctionManager.cpp:587]: The current implementation of `master` branch assumes that args[0] is float and does not do anything if it is int

File: tests/tck/features/function/round.feature (hunk 6 above)
Comment (Koukyosyumei) [reply tests/tck/features/function/round.feature:210]: I have added type mismatch cases

**Review by yixinglu — DISMISSED** [review]

**Review by wey-gu — APPROVED** [review]

**Comment by wey-gu** [comment]
@yixinglu @yixinglu I think this change is in good shape to move forward:)

**Review by yixinglu — APPROVED** [review]

File: src/common/function/FunctionManager.cpp (hunk 2 above)
Comment (czpmango) [reply src/common/function/FunctionManager.cpp:587]: `getFloat()` will crash if it is int type.

Comment (wey-gu) [reply src/common/function/FunctionManager.cpp:587]: @Koukyosyumei

File: src/common/function/FunctionManager.cpp (hunk 3 above)
Comment (czpmango) [reply src/common/function/FunctionManager.cpp:591]: You're right, my mistake.

File: src/common/function/FunctionManager.cpp (hunk 2 above)
Comment (Koukyosyumei) [reply src/common/function/FunctionManager.cpp:587]: The case condition ensures that it's float.

File: tests/tck/features/function/round.feature (hunk 5 above)
Comment (czpmango) [reply tests/tck/features/function/round.feature:206]: round() returns null if any of its input parameters are null.
```
round(1.5, null)
round(null, 1)
round(1.5, 1, null)
```

Comment (czpmango) [reply tests/tck/features/function/round.feature:206]: round() raises exception if the rounding mode is illegal string.

File: src/common/function/FunctionManager.cpp (hunk 2 above)
Comment (czpmango) [reply src/common/function/FunctionManager.cpp:587]: ok

**Review by czpmango — APPROVED** [review]
Thanks for the PR! LGTM.

**Comment by wey-gu** [comment]
Dear @Koukyosyumei ,

Many thanks for your great contribution🎉! @QingZ11 will reach out to you for the contributor certification/souvenir.

BR//Wey

**Comment by Koukyosyumei** [comment]
Thank all of you for this great opportunity and helpful suggestions! I will keep contributing to this fantastic project!!

**Pull request merged**